Report integration of testing section (#110)
anitvam/pps-bunny: doc/report.md

### Rocco

###Testing

> Sono stati inseriti dei test con il framework `ScalaTest`, con l'ausilio di `FunSpec` per rendere la loro descrizione più naturale, per testare l'implementazione delle strutture principali alla base del sistema simulato.
> Tali test consentono una verifica immediata del comportamento delle entità sviluppate anche a fronte di successive modifiche, che se effettuate nel modo sbagliato potrebbero comportare dei bug nel codice precedentemente implementato.

Reviewer: ```suggestion
Per verificare la correttezza dell'implementazione delle principali strutture dati realizzate sono stati sviluppati dei test con il framework `ScalaTest`, in particolare con l'ausilio di `FunSpec` per rendere la loro descrizione più naturale.
Tali test consentono sia una verifica immediata del comportamento delle entità sviluppate sia una verifica continua per le modifiche apportate successivamente, che se effettuate nel modo sbagliato potrebbero comportare dei bug nel codice precedentemente implementato.
```

Per la metodologia di implementazione dei test si è lasciata la libera scelta ad ogni componente del gruppo: alcuni hanno preferito l'approccio TDD (Test Driven Development) per pezzi importanti e complessi da sviluppare, mentre altri hanno scelto un approccio più tradizionale, cioè implementare prima una certa funzionalità e verificare in seguito che il comportamento di tale funzioni come atteso.  

> Sono stati inoltre introdotti test dopo lo sviluppo di più funzionalità per verificare che la loro integrazione funzioni come atteso.

Reviewer 2: ```suggestion
Sono stati inoltre introdotti test dopo lo sviluppo di un insieme di funzionalità per verificare che la loro integrazione funzioni correttamente.
```

> Si è cercato di mantenere più coverage possibile per la parte di model dell'applicazione, riuscendo a raggiungere il 100% nei package più critici e importanti del sistema come `it.unibo.pps.bunny.model.bunny` che contiene l'implementazione del coniglietto e `it.unibo.pps.bunny.model.world` che contiene la modellazione dell'ambiente di simulazione e della riproduzione.

Reviewer: okay mi ricordavo male ma su quello non abbiamo il 100%

> Si è cercato di mantenere più coverage possibile per la parte di model dell'applicazione, riuscendo a raggiungere il 100% nei package più critici e importanti del sistema come `it.unibo.pps.bunny.model.bunny` che contiene l'implementazione del coniglietto e `it.unibo.pps.bunny.model.world` che contiene la modellazione dell'ambiente di simulazione e della riproduzione.

Reviewer: Per me ci sta citare anche quel package perchè è fondamentale

> Si è cercato di mantenere più coverage possibile per la parte di model dell'applicazione, riuscendo a raggiungere il 100% nei package più critici e importanti del sistema come `it.unibo.pps.bunny.model.bunny` che contiene l'implementazione del coniglietto e `it.unibo.pps.bunny.model.world` che contiene la modellazione dell'ambiente di simulazione e della riproduzione.

Reviewer 2: Concordo

Sono di seguito riportati degli screenshot che evidenziano le informazioni sopra riportate:
![](images/scoverage_report_1.png)
![](images/scoverage_report_2.png)

> Grazie alla copertura di questi test si garantisce, grazie all'impiego della CI su Github, che sul branch `develop` non siano mai inseriti dei pezzi di codice che non funzionano correttamente, perchè se i test non passano la pull request effettuata dallo sviluppatore non viene mergiata.

Reviewer: ```suggestion
L'uso della CI su GitHub unito a tale livello di coverage garantisce che sul branch `develop` non sia mai stato inserito del codice non correttamente funzionante, infatti se i test non passano, la pull request effettuata dallo sviluppatore non viene mergiata.
```

> Per quanto riguarda invece il package `it.unibo.pps.bunny.view` non sono stati implementati molti test di `ScalaTest` per verificarne il funzionamento, in quanto la sua verifica è avvenuta prevalentemente con del beta-testing.

Reviewer: ```suggestion
Per quanto riguarda invece il package `it.unibo.pps.bunny.view` non sono stati implementati test di `ScalaTest` per verificarne il funzionamento, in quanto la sua verifica è avvenuta prevalentemente con del beta-testing, cioè eseguendo direttamente l'applicativo per verificare feature non testabili in modo automatizzato e avere una controprova visiva del corretto funzionamento delle classi di model e dell'engine.
```
